### ISSUE TITLE
Reverse-KL and JS distillation can return finite losses with NaN gradients

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#719]
## Summary

`src/soup_cli/trainer/distill.py::_compute_distill_term()` can produce non-finite
student gradients for the `reverse_kl` and `js` divergences even when all input
logits and the scalar loss are finite.

The reverse-KL path passes the differentiable student probabilities as the target
of `torch.nn.functional.kl_div`. At low, schema-valid temperatures, softmax can
underflow to zero and the target-side derivative becomes non-finite. The JS path
has the same target-side problem and additionally constructs `log_m` through a
probability-space mixture with a fixed `1e-12` clamp, which is not representable as
intended in FP16.

## Reproduction

Reproduced against current `main` at
`9f1e48b33a5fcda7e621e612ec9305cb52a38e07` with Python 3.12.2 and PyTorch
2.13.0 on CPU. The isolated test uses finite logits, temperature `0.05`, causal
shifting, and a mask with one genuinely supervised position.

| Divergence | Loss | Student gradients |
|---|---:|---|
| forward KL, FP32 | `0.5` | finite |
| reverse KL, FP32 | `0.1499999911` | **non-finite** |
| JS, FP32 | `0.00173286791` | **non-finite** |
| JS, FP16 | `NaN` | **non-finite** |

The same failure is reproduced for reverse-KL and JS in BF16. A log-space
reformulation preserves the tested loss values and returns finite gradients.

This is a kernel-level reproduction, not an end-to-end CUDA/MPS training run.

## Impact

A distillation run can report a finite forward loss while poisoning the backward
pass. The issue is reachable through a documented divergence choice and a valid
temperature; it is not an all-masked-batch case.

## Suggested fix

- Compute the distillation kernel in FP32 with `log_softmax`.
- Implement reverse KL explicitly as
  `sum(exp(log_s) * (log_s - log_t))`.
- Build the JS mixture in log space with
  `log_m = logaddexp(log_s, log_t) - log(2)`.
- Preserve the existing causal shift, trained-token mask, and `temperature ** 2`
  scaling, then cast only the final scalar as appropriate.
- Treat already non-finite input logits separately rather than hiding them.

## Acceptance criteria

- [ ] Reverse-KL and JS return finite losses and finite student gradients for the
      low-temperature reproducer in FP32, BF16, and FP16.
- [ ] Forward-KL behavior remains numerically unchanged on normal inputs.
- [ ] Tests cover causal shifting, response-only labels, padding masks, and at
      least one supervised token.
- [ ] A mutation back to the current probability-space implementation makes the
      regression test fail.

**Comment by MakazhanAlpamys** [comment]
Fixed in #736 by **@be-student**, merged as `8e86f7f0`.

`reverse_kl` returned a finite loss and a NaN gradient: `kl_div(log_t, p_s)` recomputes `log(target)` from a probability that has already underflowed to zero, so the target side is `-inf` and its derivative is NaN — on a loss curve that looks healthy, with nothing in the run to act on.

The fix is a log-space reformulation of the same objective rather than a guard on top of it. Reverse KL is now `(p_s * (log_s - log_t)).sum(-1)`, where an underflowed `p_s` multiplies a large finite `log_s` and yields the term's true limit; JS uses `logaddexp(log_s, log_t) - log(2)`, the exact identity for `log((p_s+p_t)/2)`. **One clamp was removed and none added** — the old `m.clamp(min=1e-12).log()` was itself the masking construct, and it was measurably biased: on identical distributions with a `-40` logit it returns `-5.25e-17`, a negative divergence, where the new kernel returns `0.0`.

The objective is preserved, verified rather than asserted: value and backward gradient agree with the previous kernel to fp32 round-off (rel 1.2e-07 / 1.5e-07 / 3.8e-07 for forward_kl / reverse_kl / js), and with a closed-form fp64 reference to 1.1e-08 / 1.2e-07 / 3.5e-08. Where it differs in bf16/fp16, the old kernel is the wrong one.

**Deliberately not done:** this issue's fifth suggested fix was to raise on non-finite input logits. That was implemented, then reverted at my request — non-finite logits now propagate to the loss and the gradient, which is what AMP's `GradScaler` is built to see. A 12-case test pins the propagation, so the reversal is specified tighter than the contract it replaced.

Try it:

```
soup train --config soup.yaml   # task: distill, training.distill_divergence: reverse_kl
```

The kernel returns fp32 under AMP, which `docs/training.md` now documents along with the memory cost.
